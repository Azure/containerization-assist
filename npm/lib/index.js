const { getConnection, resetConnection } = require('./client');

// Session ID counter for more readable IDs
let sessionCounter = 0;

// Generate a unique session ID
function createSession() {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
  const counter = (++sessionCounter).toString().padStart(4, '0');
  return `session-${timestamp}-${counter}`;
}

// Parse MCP response
function parseResult(mcpResult) {
  if (mcpResult?.content?.[0]?.text) {
    try {
      return JSON.parse(mcpResult.content[0].text);
    } catch (err) {
      // If it's not JSON, return as raw text
      return { raw: mcpResult.content[0].text };
    }
  }
  return mcpResult;
}

// Execute a tool with auto-session management and retry logic
async function executeTool(toolName, args = {}, retries = 1) {
  try {
    const connection = await getConnection();
    
    // Auto-generate session if not provided (except for utility tools)
    const utilityTools = ['list_tools', 'ping', 'server_status'];
    if (!args.session_id && !utilityTools.includes(toolName)) {
      args.session_id = createSession();
    }
    
    const result = await connection.callTool(toolName, args);
    return parseResult(result);
  } catch (error) {
    // Retry logic for connection failures
    if (retries > 0 && error.message.includes('Connection closed')) {
      resetConnection();
      return executeTool(toolName, args, retries - 1);
    }
    throw error;
  }
}

// ============================================
// Workflow Step Tools (10)
// ============================================

/**
 * Analyze a repository to detect language, framework, and dependencies
 * @param {string} repoPath - Path to the repository to analyze
 * @param {Object} options - Additional options (session_id, etc.)
 * @returns {Promise<Object>} Analysis results
 */
async function analyzeRepository(repoPath, options = {}) {
  if (!repoPath) {
    throw new Error('repoPath is required for analyzeRepository');
  }
  return executeTool('analyze_repository', {
    repo_path: repoPath,
    ...options
  });
}

/**
 * Generate a Dockerfile based on repository analysis
 * @param {Object} options - Generation options (session_id, base_image, etc.)
 * @returns {Promise<Object>} Dockerfile generation results
 */
async function generateDockerfile(options = {}) {
  return executeTool('generate_dockerfile', options);
}

/**
 * Build a Docker image from the generated Dockerfile
 * @param {Object} options - Build options (session_id, dockerfile, context, tags, etc.)
 * @returns {Promise<Object>} Build results with image details
 */
async function buildImage(options = {}) {
  return executeTool('build_image', options);
}

/**
 * Scan a Docker image for security vulnerabilities
 * @param {Object} options - Scan options (session_id, scanners, severity, etc.)
 * @returns {Promise<Object>} Scan results with vulnerability report
 */
async function scanImage(options = {}) {
  return executeTool('scan_image', options);
}

/**
 * Tag a Docker image with a version or label
 * @param {string} tag - The tag to apply to the image
 * @param {Object} options - Additional options (session_id, etc.)
 * @returns {Promise<Object>} Tagging results
 */
async function tagImage(tag, options = {}) {
  if (!tag) {
    throw new Error('tag is required for tagImage');
  }
  return executeTool('tag_image', {
    tag,
    ...options
  });
}

/**
 * Push a Docker image to a container registry
 * @param {string} registry - The registry URL to push to
 * @param {Object} options - Push options (session_id, credentials, etc.)
 * @returns {Promise<Object>} Push results with registry details
 */
async function pushImage(registry, options = {}) {
  if (!registry) {
    throw new Error('registry is required for pushImage');
  }
  return executeTool('push_image', {
    registry,
    ...options
  });
}

/**
 * Generate Kubernetes manifests for the application
 * @param {Object} options - Manifest options (session_id, namespace, replicas, etc.)
 * @returns {Promise<Object>} Generated manifests information
 */
async function generateK8sManifests(options = {}) {
  return executeTool('generate_k8s_manifests', options);
}

/**
 * Prepare the Kubernetes cluster for deployment
 * @param {Object} options - Cluster options (session_id, cluster_config, etc.)
 * @returns {Promise<Object>} Cluster preparation results
 */
async function prepareCluster(options = {}) {
  return executeTool('prepare_cluster', options);
}

/**
 * Deploy the application to Kubernetes
 * @param {Object} options - Deployment options (session_id, namespace, etc.)
 * @returns {Promise<Object>} Deployment results with status
 */
async function deployApplication(options = {}) {
  return executeTool('deploy_application', options);
}

/**
 * Verify the deployment is running correctly
 * @param {Object} options - Verification options (session_id, health_checks, etc.)
 * @returns {Promise<Object>} Verification results with endpoints
 */
async function verifyDeployment(options = {}) {
  return executeTool('verify_deployment', options);
}

// ============================================
// Utility Tools (3)
// ============================================

/**
 * List all available MCP tools
 * @returns {Promise<Object>} List of tools with descriptions
 */
async function listTools() {
  return executeTool('list_tools');
}

/**
 * Ping the MCP server to check connectivity
 * @returns {Promise<Object>} Ping response
 */
async function ping() {
  return executeTool('ping');
}

/**
 * Get the MCP server status
 * @returns {Promise<Object>} Server status information
 */
async function serverStatus() {
  return executeTool('server_status');
}

// ============================================
// Exports
// ============================================

module.exports = {
  // Workflow Step Tools (10)
  analyzeRepository,
  generateDockerfile,
  buildImage,
  scanImage,
  tagImage,
  pushImage,
  generateK8sManifests,
  prepareCluster,
  deployApplication,
  verifyDeployment,
  
  // Utility Tools (3)
  listTools,
  ping,
  serverStatus,
  
  // Session Management
  createSession,
  
  // Connection Management (advanced usage)
  disconnect: () => {
    const { resetConnection } = require('./client');
    resetConnection();
  }
};