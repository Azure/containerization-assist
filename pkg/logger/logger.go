package logger

import (
	"os"
	"time"

	"github.com/rs/zerolog"
)

var logger zerolog.Logger

func init() {
	output := zerolog.ConsoleWriter{
		Out:        os.Stdout,
		TimeFormat: time.RFC3339,
	}
	logger = zerolog.New(output).With().Timestamp().Logger()
	zerolog.SetGlobalLevel(zerolog.InfoLevel)
}

func Info(msg string) {
	logger.Info().Msg(msg)
}

func Infof(format string, args ...interface{}) {
	logger.Info().Msgf(format, args...)
}

func Warn(msg string) {
	logger.Warn().Msg(msg)
}

func Warnf(format string, args ...interface{}) {
	logger.Warn().Msgf(format, args...)
}

func Error(msg string) {
	logger.Error().Msg(msg)
}

func Debug(msg string) {
	logger.Debug().Msg(msg)
}
