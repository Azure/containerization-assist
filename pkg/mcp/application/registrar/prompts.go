// Package registrar handles tool and prompt registration
package registrar

import (
	"log/slog"

	"github.com/Azure/container-kit/pkg/common/errors"
	"github.com/Azure/container-kit/pkg/mcp/infrastructure/prompts"
	"github.com/mark3labs/mcp-go/server"
)

// PromptRegistrar handles prompt registration
type PromptRegistrar struct {
	logger *slog.Logger
}

// NewPromptRegistrar creates a new prompt registrar
func NewPromptRegistrar(logger *slog.Logger) *PromptRegistrar {
	return &PromptRegistrar{
		logger: logger.With("component", "prompt_registrar"),
	}
}

// RegisterAll registers all prompts with the MCP server
func (pr *PromptRegistrar) RegisterAll(mcpServer *server.MCPServer) error {
	pr.logger.Info("Registering prompts")

	// Create and register prompts
	promptRegistry := prompts.NewRegistry(mcpServer, pr.logger)
	if err := promptRegistry.RegisterAll(); err != nil {
		return errors.New(errors.CodeToolExecutionFailed, "registrar", "failed to register prompts", err)
	}

	pr.logger.Info("All prompts registered successfully")
	return nil
}
