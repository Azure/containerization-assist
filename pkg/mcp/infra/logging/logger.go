package logging

import (
	"context"
	"io"
	"log/slog"
	"sync"
	"time"
)

// Logger provides structured logging with ring buffer capability using slog.
type Logger struct {
	slogLogger *slog.Logger
	ringBuffer *RingBuffer
	config     Config
	capture    *LogCapture
	mu         sync.RWMutex
	metrics    *LogMetrics
}

// EventAdapter adapts slog operations to the Event interface.
type EventAdapter struct {
	logger  *slog.Logger
	level   slog.Level
	attrs   []slog.Attr
	message string
}

// Str adds a string field to the log event.
func (ea *EventAdapter) Str(key, value string) Event {
	ea.attrs = append(ea.attrs, slog.String(key, value))
	return ea
}

// Err adds an error field to the log event.
func (ea *EventAdapter) Err(err error) Event {
	if err != nil {
		ea.attrs = append(ea.attrs, slog.Any("error", err))
	}
	return ea
}

// Int adds an integer field to the log event.
func (ea *EventAdapter) Int(key string, value int) Event {
	ea.attrs = append(ea.attrs, slog.Int(key, value))
	return ea
}

// Float64 adds a float64 field to the log event.
func (ea *EventAdapter) Float64(key string, value float64) Event {
	ea.attrs = append(ea.attrs, slog.Float64(key, value))
	return ea
}

// Bool adds a boolean field to the log event.
func (ea *EventAdapter) Bool(key string, value bool) Event {
	ea.attrs = append(ea.attrs, slog.Bool(key, value))
	return ea
}

// Msg logs the event with the given message.
func (ea *EventAdapter) Msg(msg string) {
	ea.logger.LogAttrs(context.Background(), ea.level, msg, ea.attrs...)
}

// Event defines the interface for log events.
type Event interface {
	Str(key, value string) Event
	Err(err error) Event
	Int(key string, value int) Event
	Float64(key string, value float64) Event
	Bool(key string, value bool) Event
	Msg(msg string)
}

// LogCapture provides log capture functionality with ring buffer storage.
type LogCapture struct {
	ringBuffer *RingBuffer
	config     LogCaptureConfig
	active     bool
}

// LogCaptureConfig defines log capture configuration.
type LogCaptureConfig struct {
	MaxEntries    int           `json:"max_entries"`
	MinLevel      Level         `json:"min_level"`
	RetentionTime time.Duration `json:"retention_time"`
	BufferSize    int           `json:"buffer_size"`
}

// LogMetrics tracks logging performance metrics.
type LogMetrics struct {
	TotalLogs   int64         `json:"total_logs"`
	ErrorLogs   int64         `json:"error_logs"`
	WarningLogs int64         `json:"warning_logs"`
	DebugLogs   int64         `json:"debug_logs"`
	InfoLogs    int64         `json:"info_logs"`
	AverageTime time.Duration `json:"average_time"`
	LastLogTime time.Time     `json:"last_log_time"`
}

// LogEntry represents a captured log entry.
type LogEntry struct {
	Timestamp time.Time
	Level     Level
	Message   string
	Fields    map[string]interface{}
	Caller    string
	Error     error
}

// UnifiedLogger is an alias for Logger to match the public API
type UnifiedLogger = Logger

// NewLogger creates a new unified logger with the specified configuration.
func NewLogger(config Config) *Logger {
	var ringBuffer *RingBuffer
	var logCapture *LogCapture
	var output io.Writer = config.Output

	// Initialize ring buffer if enabled
	if config.EnableRingBuffer {
		ringBuffer = NewRingBuffer(config.BufferSize)
		// Create multi-writer (output + ring buffer)
		output = io.MultiWriter(config.Output, ringBuffer)
	}

	// Initialize log capture if enabled
	if config.EnableRingBuffer {
		captureConfig := LogCaptureConfig{
			MaxEntries:    config.BufferSize,
			MinLevel:      config.Level,
			RetentionTime: 24 * time.Hour,
			BufferSize:    config.BufferSize,
		}

		logCapture = &LogCapture{
			ringBuffer: ringBuffer,
			config:     captureConfig,
			active:     true,
		}
	}

	// Configure slog
	opts := &slog.HandlerOptions{
		Level:     toSlogLevel(config.Level),
		AddSource: config.EnableCaller,
	}

	var handler slog.Handler
	if config.EnableStructuredLogging {
		handler = slog.NewJSONHandler(output, opts)
	} else {
		handler = slog.NewTextHandler(output, opts)
	}

	// Add default fields
	if len(config.Fields) > 0 {
		attrs := make([]slog.Attr, 0, len(config.Fields))
		for k, v := range config.Fields {
			attrs = append(attrs, slog.Any(k, v))
		}
		handler = handler.WithAttrs(attrs)
	}

	slogLogger := slog.New(handler)

	// Initialize metrics
	metrics := &LogMetrics{
		LastLogTime: time.Now(),
	}

	return &Logger{
		slogLogger: slogLogger,
		ringBuffer: ringBuffer,
		capture:    logCapture,
		config:     config,
		metrics:    metrics,
	}
}

// toSlogLevel converts our Level to slog.Level.
func toSlogLevel(l Level) slog.Level {
	switch l {
	case LevelDebug:
		return slog.LevelDebug
	case LevelInfo:
		return slog.LevelInfo
	case LevelWarn:
		return slog.LevelWarn
	case LevelError:
		return slog.LevelError
	case LevelFatal:
		return slog.LevelError + 4 // slog doesn't have fatal, use higher than error
	default:
		return slog.LevelInfo
	}
}

// GetRecentLogs returns recent log entries from the ring buffer.
func (l *Logger) GetRecentLogs() []LogEntry {
	if l.ringBuffer == nil {
		return nil
	}

	l.mu.RLock()
	defer l.mu.RUnlock()
	return l.ringBuffer.GetAll()
}

// GetLogsSince returns log entries since the specified time.
func (l *Logger) GetLogsSince(since time.Time) []LogEntry {
	if l.ringBuffer == nil {
		return nil
	}

	l.mu.RLock()
	defer l.mu.RUnlock()

	allEntries := l.ringBuffer.GetAll()
	var result []LogEntry

	for _, entry := range allEntries {
		if entry.Timestamp.After(since) {
			result = append(result, entry)
		}
	}

	return result
}

// GetLogsByLevel returns log entries with the specified level.
func (l *Logger) GetLogsByLevel(level Level) []LogEntry {
	if l.ringBuffer == nil {
		return nil
	}

	l.mu.RLock()
	defer l.mu.RUnlock()

	allEntries := l.ringBuffer.GetAll()
	var result []LogEntry

	for _, entry := range allEntries {
		if entry.Level == level {
			result = append(result, entry)
		}
	}

	return result
}

// Clear clears the captured logs.
func (l *Logger) Clear() {
	if l.ringBuffer == nil {
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()
	l.ringBuffer.Clear()
}

// Size returns the number of captured log entries.
func (l *Logger) Size() int {
	if l.ringBuffer == nil {
		return 0
	}

	l.mu.RLock()
	defer l.mu.RUnlock()
	return l.ringBuffer.Size()
}

// WithField adds a structured field to the logger context.
func (l *Logger) WithField(key string, value interface{}) Standards {
	return &Logger{
		slogLogger: l.slogLogger.With(key, value),
		ringBuffer: l.ringBuffer,
		capture:    l.capture,
		config:     l.config,
		metrics:    l.metrics,
	}
}

// WithFields adds multiple fields to the logger context.
func (l *Logger) WithFields(fields map[string]interface{}) Standards {
	args := make([]any, 0, len(fields)*2)
	for k, v := range fields {
		args = append(args, k, v)
	}

	return &Logger{
		slogLogger: l.slogLogger.With(args...),
		ringBuffer: l.ringBuffer,
		capture:    l.capture,
		config:     l.config,
		metrics:    l.metrics,
	}
}

// WithError adds an error field to the logger context.
func (l *Logger) WithError(err error) Standards {
	return &Logger{
		slogLogger: l.slogLogger.With("error", err),
		ringBuffer: l.ringBuffer,
		capture:    l.capture,
		config:     l.config,
		metrics:    l.metrics,
	}
}

// WithComponent adds a component name to the logger context.
func (l *Logger) WithComponent(component string) Standards {
	return l.WithField("component", component)
}

// WithTraceID adds a trace ID to the logger context.
func (l *Logger) WithTraceID(traceID string) Standards {
	return l.WithField("trace_id", traceID)
}

// WithSpanID adds a span ID to the logger context.
func (l *Logger) WithSpanID(spanID string) Standards {
	return l.WithField("span_id", spanID)
}

// WithRequestID adds a request ID to the logger context.
func (l *Logger) WithRequestID(requestID string) Standards {
	return l.WithField("request_id", requestID)
}

// WithUserID adds a user ID to the logger context.
func (l *Logger) WithUserID(userID string) Standards {
	return l.WithField("user_id", userID)
}

// Info returns a log event for info level logging.
func (l *Logger) Info() Event {
	l.updateMetrics(LevelInfo)
	return &EventAdapter{
		logger: l.slogLogger,
		level:  slog.LevelInfo,
		attrs:  []slog.Attr{},
	}
}

// Error returns a log event for error level logging.
func (l *Logger) Error() Event {
	l.updateMetrics(LevelError)
	return &EventAdapter{
		logger: l.slogLogger,
		level:  slog.LevelError,
		attrs:  []slog.Attr{},
	}
}

// Debug returns a log event for debug level logging.
func (l *Logger) Debug() Event {
	l.updateMetrics(LevelDebug)
	return &EventAdapter{
		logger: l.slogLogger,
		level:  slog.LevelDebug,
		attrs:  []slog.Attr{},
	}
}

// Warn returns a log event for warning level logging.
func (l *Logger) Warn() Event {
	l.updateMetrics(LevelWarn)
	return &EventAdapter{
		logger: l.slogLogger,
		level:  slog.LevelWarn,
		attrs:  []slog.Attr{},
	}
}

// updateMetrics updates internal logging metrics.
func (l *Logger) updateMetrics(level Level) {
	if l.metrics == nil {
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	l.metrics.TotalLogs++
	l.metrics.LastLogTime = time.Now()

	switch level {
	case LevelDebug:
		l.metrics.DebugLogs++
	case LevelInfo:
		l.metrics.InfoLogs++
	case LevelWarn:
		l.metrics.WarningLogs++
	case LevelError:
		l.metrics.ErrorLogs++
	}
}

// GetMetrics returns current logging metrics.
func (l *Logger) GetMetrics() LogMetrics {
	if l.metrics == nil {
		return LogMetrics{}
	}

	l.mu.RLock()
	defer l.mu.RUnlock()
	return *l.metrics
}

// Counter logs a counter metric.
func (l *Logger) Counter(name string, value int64, tags map[string]string) {
	attrs := []slog.Attr{
		slog.String("metric_type", "counter"),
		slog.String("metric_name", name),
		slog.Int64("metric_value", value),
	}

	for k, v := range tags {
		attrs = append(attrs, slog.String("tag_"+k, v))
	}

	l.slogLogger.LogAttrs(context.Background(), slog.LevelInfo, "metric recorded", attrs...)
}

// Gauge logs a gauge metric.
func (l *Logger) Gauge(name string, value float64, tags map[string]string) {
	attrs := []slog.Attr{
		slog.String("metric_type", "gauge"),
		slog.String("metric_name", name),
		slog.Float64("metric_value", value),
	}

	for k, v := range tags {
		attrs = append(attrs, slog.String("tag_"+k, v))
	}

	l.slogLogger.LogAttrs(context.Background(), slog.LevelInfo, "metric recorded", attrs...)
}

// Timer logs a timer metric.
func (l *Logger) Timer(name string, duration time.Duration, tags map[string]string) {
	attrs := []slog.Attr{
		slog.String("metric_type", "timer"),
		slog.String("metric_name", name),
		slog.Duration("metric_value", duration),
	}

	for k, v := range tags {
		attrs = append(attrs, slog.String("tag_"+k, v))
	}

	l.slogLogger.LogAttrs(context.Background(), slog.LevelInfo, "metric recorded", attrs...)
}

// Histogram logs a histogram metric.
func (l *Logger) Histogram(name string, value float64, tags map[string]string) {
	attrs := []slog.Attr{
		slog.String("metric_type", "histogram"),
		slog.String("metric_name", name),
		slog.Float64("metric_value", value),
	}

	for k, v := range tags {
		attrs = append(attrs, slog.String("tag_"+k, v))
	}

	l.slogLogger.LogAttrs(context.Background(), slog.LevelInfo, "metric recorded", attrs...)
}
