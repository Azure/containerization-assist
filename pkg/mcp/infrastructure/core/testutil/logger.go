// Package testutil provides common testing utilities for the MCP codebase.
// It includes helpers for logging, mocking, fixtures, and assertions.
package testutil

import (
	"bytes"
	"log/slog"
	"testing"
)

// TestLogger represents a logger for testing that captures output
type TestLogger struct {
	*slog.Logger
	Buffer *bytes.Buffer
}

// NewTestLogger creates a new test logger that captures all output
func NewTestLogger(t *testing.T) *TestLogger {
	t.Helper()

	buffer := &bytes.Buffer{}
	handler := slog.NewTextHandler(buffer, &slog.HandlerOptions{
		Level: slog.LevelDebug, // Capture all levels in tests
	})

	logger := slog.New(handler).With("test", t.Name())

	return &TestLogger{
		Logger: logger,
		Buffer: buffer,
	}
}

// NewDiscardLogger creates a logger that discards all output
func NewDiscardLogger() *slog.Logger {
	return slog.New(slog.NewTextHandler(&bytes.Buffer{}, nil))
}

// GetOutput returns the captured log output as a string
func (tl *TestLogger) GetOutput() string {
	return tl.Buffer.String()
}

// Contains checks if the log output contains a specific string
func (tl *TestLogger) Contains(substr string) bool {
	return bytes.Contains(tl.Buffer.Bytes(), []byte(substr))
}

// Reset clears the captured log output
func (tl *TestLogger) Reset() {
	tl.Buffer.Reset()
}

// AssertLogged checks if a message was logged and fails the test if not
func AssertLogged(t *testing.T, logger *TestLogger, message string) {
	t.Helper()

	if !logger.Contains(message) {
		t.Errorf("Expected log message not found: %q\nActual output:\n%s", message, logger.GetOutput())
	}
}

// AssertNotLogged checks if a message was NOT logged and fails the test if it was
func AssertNotLogged(t *testing.T, logger *TestLogger, message string) {
	t.Helper()

	if logger.Contains(message) {
		t.Errorf("Unexpected log message found: %q\nActual output:\n%s", message, logger.GetOutput())
	}
}
