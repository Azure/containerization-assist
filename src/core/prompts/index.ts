/**
 * Core Prompts Module
 *
 * Exports the prompt registry and related types for managing
 * statically imported JSON prompt templates.
 */

export { PromptRegistry } from './registry';
export { StaticPromptLoader, type PromptFile, type ParameterSpec } from './static-loader';
