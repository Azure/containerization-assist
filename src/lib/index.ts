/**
 * Lib Utilities Barrel File
 *
 * Clean, focused exports for pure utility functions.
 * Organized by domain for easy navigation.
 */

// Content processing utilities
export {
  extractJsonFromText,
  sanitizeResponseText,
  estimateTokenCount,
  truncatePreservingStructure,
  extractStructuredContent,
  normalizeWhitespace,
  appearsToBeCode,
} from './content-utils';

// Capability management
export {
  CapabilitySet,
  PredefinedCapabilities,
  createCapabilitySet,
  mergeCapabilities,
  canHandleTask,
  findCapableTools,
  getRecommendedCapabilities,
} from './capabilities';

// Text processing (legacy - prefer content-utils for new code)
export {
  stripFencesAndNoise,
  isValidDockerfileContent,
  extractBaseImage,
  isValidKubernetesContent,
} from './text-processing';

// Core utilities
export { createLogger } from './logger';
export { extractErrorMessage } from './error-utils';

// Async utilities
export { type Cache } from './cache';

// Regular expression patterns
export { DOCKERFILE_FENCE, YAML_FENCE, GENERIC_FENCE, AS_CLAUSE } from './regex-patterns';
