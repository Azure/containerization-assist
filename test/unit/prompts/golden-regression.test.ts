/**
 * Golden File Regression Tests for Critical Prompts
 *
 * Tests prompt rendering with known parameters and snapshots results
 * to ensure no functionality is lost during format changes.
 */

import { initializeRegistry, getPrompt } from '../../../src/prompts/prompt-registry';
import { join } from 'path';
import { createLogger } from '../../../src/lib/logger';

describe('Prompt Golden File Regression Tests', () => {
  const logger = createLogger({ name: 'test', level: 'silent' });

  beforeAll(async () => {
    // Initialize registry with actual prompts directory
    const promptsDir = join(process.cwd(), 'src/prompts');
    const result = await initializeRegistry(promptsDir, logger);

    if (!result.ok) {
      throw new Error(`Failed to initialize prompts: ${result.error}`);
    }
  });

  describe('deploy-application prompt', () => {
    const testParams = {
      manifests: ['deployment.yaml', 'service.yaml'],
      namespace: 'production',
      cluster: 'main-cluster',
      environment: 'production',
      strategy: 'rolling',
      existingDeployment: {
        name: 'my-app',
        replicas: 3,
        version: '1.2.0'
      },
      sessionAnalysis: {
        risks: ['high-traffic'],
        recommendations: ['enable-hpa']
      },
      knowledge: 'Best practices for production deployments',
      policy: 'Security-first deployment policy'
    };

    it('should render consistently with known parameters', async () => {
      const result = await getPrompt('deploy-application', testParams);

      expect(result.ok).toBe(true);
      if (result.ok) {
        const content = result.value.content;

        // Verify key sections are present
        expect(content).toContain('Deployment Configuration:');
        expect(content).toContain('production'); // namespace
        expect(content).toContain('deployment.yaml'); // manifests
        expect(content).toContain('main-cluster'); // cluster
        expect(content).toContain('rolling'); // strategy
        expect(content).toContain('"name": "my-app"'); // existing deployment (as JSON)
        expect(content).toContain('"risks"'); // session analysis (as JSON)
        expect(content).toContain('Best practices'); // knowledge
        expect(content).toContain('Security-first'); // policy

        // Verify JSON structure request is present
        expect(content).toContain('"deploymentStrategy"');
        expect(content).toContain('"preDeploymentChecks"');
        expect(content).toContain('"rolloutSteps"');
        expect(content).toContain('"healthChecks"');
        expect(content).toContain('"riskAssessment"');

        // Snapshot test - this will create/verify golden file
        expect(content).toMatchSnapshot('deploy-application-golden');
      }
    });

    it('should handle minimal parameters correctly', async () => {
      const minimalParams = {
        manifests: ['app.yaml'],
        namespace: 'default'
      };

      const result = await getPrompt('deploy-application', minimalParams);

      expect(result.ok).toBe(true);
      if (result.ok) {
        const content = result.value.content;

        // Should contain provided params
        expect(content).toContain('app.yaml');
        expect(content).toContain('default');

        // Optional params should be empty or handled gracefully
        // Just verify the content was generated without errors

        // Structure should still be intact
        expect(content).toContain('Deployment Configuration:');
        expect(content).toContain('"deploymentStrategy"');
      }
    });
  });

  describe('generate-k8s-manifests prompt', () => {
    it('should be loadable and render with test parameters', async () => {
      const testParams = {
        appName: 'test-app',
        imageId: 'nginx:1.21', // Use imageId as required by the schema
        port: 80,
        environment: 'staging',
        replicas: 2
      };

      const result = await getPrompt('generate-k8s-manifests', testParams);

      expect(result.ok).toBe(true);
      if (result.ok) {
        expect(result.value.content).toContain('test-app');
        expect(result.value.content).toContain('nginx:1.21');
        expect(result.value.content.length).toBeGreaterThan(100);
      }
    });
  });

  describe('nested variable handling', () => {
    it('should handle complex nested objects correctly', () => {
      const complexParams = {
        app: {
          name: 'complex-app',
          config: {
            database: {
              host: 'db.example.com',
              port: 5432
            }
          }
        }
      };

      // Test with a template that uses nested variables
      const testTemplate = 'App: {{app.name}}, DB: {{app.config.database.host}}:{{app.config.database.port}}';

      // Since we can't easily test this with an actual prompt, we'll verify
      // the template rendering logic works correctly using a simple helper
      function renderTemplate(template: string, params: Record<string, unknown>): string {
        return template.replace(/\{\{\s*(\w+(?:\.\w+)*)\s*\}\}/g, (_, path) => {
          const value = path.split('.').reduce((current: any, key: string) => {
            return current && typeof current === 'object' ? current[key] : undefined;
          }, params);
          return value !== undefined ? String(value) : '';
        });
      }

      const rendered = renderTemplate(testTemplate, complexParams);
      expect(rendered).toBe('App: complex-app, DB: db.example.com:5432');
    });
  });

  describe('outputFormat: json handling', () => {
    it('should preserve JSON output format instructions', async () => {
      const result = await getPrompt('deploy-application', {
        manifests: ['test.yaml'],
        namespace: 'test'
      });

      if (result.ok) {
        const content = result.value.content;

        // Should contain JSON format instruction
        expect(content).toContain('```json');
        expect(content).toContain('```');

        // Should have proper JSON structure examples
        expect(content).toMatch(/"[\w]+": "[^"]*"/); // JSON key-value pairs
      }
    });
  });
});