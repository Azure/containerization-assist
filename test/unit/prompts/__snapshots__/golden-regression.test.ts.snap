// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt Golden File Regression Tests deploy-application prompt should render consistently with known parameters: deploy-application-golden 1`] = `
"Analyze and provide deployment strategy recommendations for a Kubernetes deployment.

Deployment Configuration:
- Manifests: deployment.yaml,service.yaml
- Namespace: production

- Cluster: main-cluster


- Environment: production


- Preferred Strategy: rolling



Existing Deployment:
{
  "name": "my-app",
  "replicas": 3,
  "version": "1.2.0"
}



Previous Analysis:
{
  "risks": [
    "high-traffic"
  ],
  "recommendations": [
    "enable-hpa"
  ]
}



Deployment Best Practices:
Best practices for production deployments



Deployment Policies:
Security-first deployment policy


Provide a JSON response with the following structure:
\`\`\`json
{
  "deploymentStrategy": {
    "type": "string (rolling|recreate|blue-green|canary)",
    "reasoning": "string",
    "parameters": {
      "maxSurge": "string",
      "maxUnavailable": "string",
      "progressDeadlineSeconds": number
    }
  },
  "preDeploymentChecks": [
    {
      "check": "string",
      "required": boolean,
      "description": "string"
    }
  ],
  "rolloutSteps": [
    {
      "step": number,
      "action": "string",
      "validation": "string"
    }
  ],
  "healthChecks": {
    "readinessProbe": {
      "enabled": boolean,
      "initialDelaySeconds": number,
      "periodSeconds": number
    },
    "livenessProbe": {
      "enabled": boolean,
      "initialDelaySeconds": number,
      "periodSeconds": number
    }
  },
  "scaling": {
    "replicas": number,
    "autoScaling": boolean,
    "minReplicas": number,
    "maxReplicas": number
  },
  "monitoring": {
    "metrics": ["string"],
    "alerts": ["string"],
    "dashboards": ["string"]
  },
  "rollbackTriggers": ["string"],
  "postDeploymentValidation": ["string"],
  "estimatedDeploymentTime": "string",
  "riskAssessment": {
    "level": "string (low|medium|high)",
    "factors": ["string"],
    "mitigations": ["string"]
  }
}
\`\`\`

Focus on safe, reliable deployment strategies appropriate for the target environment.
"
`;
